py/dml/crep.py

```diff
@@ -19,8 +19,52 @@
     'ctype',
     'conf_obj',
     'cloggroup',
+    'dev',
+    'require_dev',
+    'maybe_dev',
+    'maybe_dev_arg',
+    'StaticContext',
+    'TypedParamContext',
     )
 
+
+class StaticContext:
+    active = False
+
+    def __enter__(self):
+        self.prev_context = StaticContext.active
+        StaticContext.active = True
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        StaticContext.active = self.prev_context
+
+class TypedParamContext(StaticContext):
+    active = False
+
+    def __enter__(self):
+        StaticContext.__enter__(self)
+        self.prev_typed_context = TypedParamContext.active
+        TypedParamContext.active = True
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        StaticContext.__exit__(self, exc_type, exc_val, exc_tb)
+        TypedParamContext.active = self.prev_typed_context
+
+def dev(site):
+    require_dev(site)
+    return "_dev"
+
+def require_dev(site):
+    if StaticContext.active:
+        raise EINDEPENDENTVIOL(site)
+
+def maybe_dev(independent):
+    return [] if independent else ['_dev']
```

Comment:
`else [dev(site)]`? The function name suggests that responsibilities are similar to `dev()`.

Reply:
No, it's different. The name should be changed, not a `require_dev` check -- the latter would break it.

Reply:
Actually, I have a mind to scrap this function and replace calls to it with `['_dev'] * independent` -- it's about as clean as a function call.
Still, that can't be done with `maybe_dev_arg`, and I can't for the life of me figure out how to rename it.